Add graceful period to server (#499)

arielmorelli opened the pull request:
<!-- 
Please use this template when submitting a new feature request with as much details as possible. Not doing so may result in the issue not being addressed in a timely manner or eventually closed.
-->
# Title

Add graceful period to server (#499)

## Summary 

Add graceful period before stopping the server.

## Type

- [ ] Bug fix
- [x] Feature request
- [x] Enhancement
- [ ] Docs

## Why?

This is important when scaling up/down the application. 

## Requirements

?

## How to try it?

Use a long-running endpoint (like `sleep(10)`. Kill gogatekeeper before the route is done.

## Documentation

<!-- 
Link to the documentation pull-request if necessary
-->

## Additional Information

<!-- 
Any additional information that you believe worth mentioning
-->

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.

p53 commented on docs/content/configuration/_index.md:
```diff
@@ -124,6 +124,7 @@ weight: 2
 |    --enabled-proxy-protocol                 | enable proxy protocol | false | PROXY_ENABLE_PROXY_PROTOCOL
 |    --max-idle-connections value             | max idle upstream / keycloak connections to keep alive, ready for reuse | 0 | PROXY_MAX_IDLE_CONNS
 |    --max-idle-connections-per-host value    | limits the number of idle connections maintained per host | 0 | PROXY_MAX_IDLE_CONNS_PER_HOST
+|    --server-grace-timeout value             | the server graceful period before shutdown | 10s | PROXY_SERVER_READ_TIMEOUT
```
you have here PROXY_SERVER_READ_TIMEOUT, which is not right

arielmorelli replied:
Opsy! Fixed

p53 commented on pkg/keycloak/proxy/server.go:
```diff
@@ -1006,6 +1006,14 @@ func (r *OauthProxy) Run() error {
+func (r *OauthProxy) Shutdown() {
+	gracefulPeriod := r.Config.ServerGraceTimeout
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(time.Millisecond*gracefulPeriod))
+    defer cancel()
```
here is bad indentation

arielmorelli replied:
Fixed

p53 commented on pkg/keycloak/proxy/server.go:
```diff
@@ -1006,6 +1006,14 @@ func (r *OauthProxy) Run() error {
+// Shutdown finishes the proxy service with gracefully period
+func (r *OauthProxy) Shutdown() {
+	gracefulPeriod := r.Config.ServerGraceTimeout
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(time.Millisecond*gracefulPeriod))
```
hmm gracefulPeriod is in seconds why we are multiplying it with miliseconds?

arielmorelli replied:
I was following the other durations. Set as Seconds

p53 reviewed (commented):
maybe if you would add also tests, but they are a little bit cumbersome to setup, i can add something later

p53 commented on pkg/keycloak/proxy/server.go:
```diff
@@ -1006,6 +1006,14 @@ func (r *OauthProxy) Run() error {
+// Shutdown finishes the proxy service with gracefully period
+func (r *OauthProxy) Shutdown() {
+	gracefulPeriod := r.Config.ServerGraceTimeout
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(time.Second*gracefulPeriod))
```
there should not be time.Duration() only gracefulPeriod as it is already variable of type time.Duration

arielmorelli replied:
My bad! Fixed.

p53 commented:
@arielmorelli, see https://github.com/gogatekeeper/gatekeeper/actions/runs/10686935891/job/29625010772?pr=500#step:4:27 (failed Lint)

arielmorelli commented:
I'm so rusted in go!

Is there any way to build locally using docker? I see that depends on github specific to run

arielmorelli commented:
@p53 can you check if I've added the tests correctly?

p53 commented:
@arielmorelli you can build image using simply `docker buildx build --platform linux/amd64 .`

p53 commented:
> @p53 can you check if I've added the tests correctly?

Yup, they look nice from functional point of view, a little bit from organizational point, they test same feature and they have similar setup so it would be good to group them in table test way (haven't migrated all tests to this kind of setup, but quite many yes), e.g. https://github.com/gogatekeeper/gatekeeper/blob/13590c5b2b3aee2a2647d81043b5b7065e55e59a/pkg/testsuite/server_test.go#L114

arielmorelli commented:
> > @p53 can you check if I've added the tests correctly?
> 
> Yup, they look nice from functional point of view, a little bit from organizational point, they test same feature and they have similar setup so it would be good to group them in table test way (haven't migrated all tests to this kind of setup, but quite many yes), e.g.
> 
> https://github.com/gogatekeeper/gatekeeper/blob/13590c5b2b3aee2a2647d81043b5b7065e55e59a/pkg/testsuite/server_test.go#L114

Got the idea. Added as a list of tests and fixed linter.

There is any easy way to run `golangci-lint` locally like in the pipeline?

arielmorelli commented:
> @arielmorelli if you have it locally just run golangci-lint run . (you have to have same version of golanci-lint as in pipeline, usually those warnings/errors differ quite a lot between versions)

worked fine with `golangci-lint run ./...` but I can see a lot of issues that are not related to my PR. I guess is ready now :)

p53 commented on pkg/testsuite/server_test.go:
```diff
@@ -42,6 +43,8 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+const ForwardingGrantTypePassword = "password"
```
only one last small thing, can you move this to constant.go file in this testsuite package, thx!

arielmorelli replied:
> only one last small thing, can you move this to constant.go file in this testsuite package, thx!

moved

p53 commented:
> > @arielmorelli if you have it locally just run golangci-lint run . (you have to have same version of golanci-lint as in pipeline, usually those warnings/errors differ quite a lot between versions)
> 
> worked fine with `golangci-lint run ./...` but I can see a lot of issues that are not related to my PR. I guess is ready now :)

sorry yes, right now it needs to be run with `./...`, was changing directory structure...anyway are you using same version of golangci-lint? Sometimes e.g vscodium install golang tools along with golangci-lint so then you might have version installed by vscode and also version which you installed on your own. If you are on unix, check by whereis command, check version of each and run it with full path. I run now golangci-lint localy and it outputs same output as in github action, as expected

p53 commented on pkg/testsuite/server_test.go:
```diff
@@ -2166,6 +2167,87 @@ func TestCustomHTTPMethod(t *testing.T) {
+					defer waitShutdown.Done()
+					<-time.After(time.Duration(200) * time.Millisecond)
+					if err := proxy.Shutdown(); err != nil {
+						failedToShutdownServer = true
```
use t.Error instead t.Fatal and that should do it, communicating with goroutines is better to be done with channels, using variables like this can introduce data race problems

arielmorelli replied:
I said my golang skills are rusted :sweat_smile: 

fixed

arielmorelli commented:
> > > @arielmorelli if you have it locally just run golangci-lint run . (you have to have same version of golanci-lint as in pipeline, usually those warnings/errors differ quite a lot between versions)
> > 
> > 
> > worked fine with `golangci-lint run ./...` but I can see a lot of issues that are not related to my PR. I guess is ready now :)
> 
> sorry yes, right now it needs to be run with `./...`, was changing directory structure...anyway are you using same version of golangci-lint? Sometimes e.g vscodium install golang tools along with golangci-lint so then you might have version installed by vscode and also version which you installed on your own. If you are on unix, check by whereis command, check version of each and run it with full path. I run now golangci-lint localy and it outputs same output as in github action, as expected
> 
> ```
> whereis golangci-lint
> golangci-lint: /usr/bin/golangci-lint /bin/golangci-lint
> 
> /bin/golangci-lint --version
> golangci-lint has version v1.59.1 built with go1.22.4 from (unknown, modified: ?, mod sum: "h1:CRRLu1JbhK5avLABFJ/OHVSQ0Ie5c4ulsOId1h3TTks=") on (unknown)
> 
> /bin/golangci-lint run ./...
> WARN [config_reader] The configuration option `linters.govet.check-shadowing` is deprecated. Please enable `shadow` instead, if you are not using `enable-all`. 
> WARN [lintersdb] The linter "exhaustivestruct" is deprecated (step 2) and deactivated. It should be removed from the list of disabled linters. https://golangci-lint.run/product/roadmap/#linter-deprecation-cycle 
> WARN [lintersdb] The linter "maligned" is deprecated (step 2) and deactivated. It should be removed from the list of disabled linters. https://golangci-lint.run/product/roadmap/#linter-deprecation-cycle 
> WARN [lintersdb] The name "goerr113" is deprecated. The linter has been renamed to: err113. 
> WARN The linter 'execinquery' is deprecated (since v1.58.0) due to: The repository of the linter has been archived by the owner.  
> pkg/testsuite/server_test.go:2248:1: unnecessary trailing newline (whitespace)
> 
> ^
> pkg/testsuite/server_test.go:2239:7: testinggoroutine: call to (*testing.T).Fatal from a non-test goroutine (govet)
>                                                 t.Fatal(err)
>                                                 ^
> pkg/testsuite/server_test.go:2235:5: SA2002: the goroutine calls T.Fatal, which must be called in the same goroutine as the test (staticcheck)
>                                 go func() {
>                                 ^
> pkg/testsuite/server_test.go:2184:2: variable name 's' is too short for the scope of its usage (varnamelen)
>         s := httptest.NewServer(&FakeUpstreamService{})
>         ^
> ```

Yeah, to avoid the whole habit whole of installing stuff, running on docker is easier: `docker run -t --rm -v $(pwd):/app -w /app golangci/golangci-lint:v1.59.1 golangci-lint run -v`


fixed all linter issues.

p53 commented:
@arielmorelli thanks for your work!
